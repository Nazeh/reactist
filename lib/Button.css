.reactist_button {
  font-size: 0.875rem;
  color: white;
  font-weight: 400;
  line-height: 1.7;
  border: 1px solid rgba(0, 0, 0, 0.1);
  border-radius: 3px;
  box-sizing: border-box;
  background-color: #3f82ef;
  padding: 5px 15px;
  cursor: pointer;
}
.reactist_button .wrapper {
  display: inline-block;
}
.reactist_button span {
  user-select: none;
}
.reactist_button:disabled,
.reactist_button.busy {
  opacity: 0.4;
  cursor: not-allowed;
}
.reactist_button:not(.secondary):not(.white):not(.danger):hover {
  background-color: #3b7be2;
}
.reactist_button.small {
  font-size: 0.81rem;
  color: white;
  font-weight: 400;
  line-height: 1.6;
  padding: 5px 10px;
}
.reactist_button.large {
  padding: 10px 15px;
}
.reactist_button.danger {
  background-color: #de4c4a;
}
.reactist_button.danger:not([disabled]):not(.busy):hover {
  background-color: #cf2826;
}
.reactist_button.secondary {
  background-color: white;
  border: 1px solid #dcdcdc;
  border-radius: 3px;
  color: #202020;
}
.reactist_button.secondary:not([disabled]):not(.busy):hover {
  background-color: #f9f9f9;
}
.reactist_button.white {
  background-color: white;
  border: 1px solid #3f82ef;
  border-radius: 3px;
  color: #3f82ef;
}
.reactist_button.white:not([disabled]):not(.busy):hover {
  background-color: #e7f3fd;
}
.reactist_button.busy {
  cursor: progress !important;
}
.reactist_button.busy.white,
.reactist_button.busy.secondary {
  border: 1px solid #dcdcdc;
  border-radius: 3px;
  background-color: #dcdcdc;
  color: #808080;
}
.reactist_button.busy:after {
  background-repeat: no-repeat;
  background-size: 15px;
  content: '';
  display: inline-block;
  height: 15px;
  margin-left: 12px;
  vertical-align: middle;
  width: 15px;
  animation-duration: 1s;
  animation-iteration-count: infinite;
  animation-name: reactistRotateRight;
  animation-timing-function: linear;
  background-image: url("data:image/svg+xml,%3Csvg version='1.1' id='Layer_1' xmlns='http://www.w3.org/2000/svg' xmlns:xlink='http://www.w3.org/1999/xlink' x='0px' y='0px' width='16.257px' height='16.257px' viewBox='-147.813 206.75 16.257 16.257' enable-background='new -147.813 206.75 16.257 16.257' xml:space='preserve'%3E %3Cg id='Symbols'%3E %3Cg id='Loading-spinner' transform='translate(0.000000, -2.000000)'%3E %3Cg id='Group'%3E %3Cg id='Rectangle-12'%3E %3C/g%3E %3Cdefs%3E %3Cfilter id='Adobe_OpacityMaskFilter' filterUnits='userSpaceOnUse' x='-147.684' y='210.929' width='16' height='13.95'%3E %3CfeColorMatrix type='matrix' values='1 0 0 0 0 0 1 0 0 0 0 0 1 0 0 0 0 0 1 0'/%3E %3C/filter%3E %3C/defs%3E %3Cmask maskUnits='userSpaceOnUse' x='-147.684' y='210.929' width='16' height='13.95' id='mask-2_1_'%3E %3Cg filter='url(%23Adobe_OpacityMaskFilter)'%3E %3Crect id='path-1_1_' x='-148.584' y='207.979' fill='%23FFFFFF' width='18' height='18'/%3E %3C/g%3E %3C/mask%3E %3Cpath id='Oval' mask='url(%23mask-2_1_)' fill='none' stroke='%23FFFFFF' stroke-width='2' stroke-linecap='round' stroke-linejoin='round' d=' M-144.634,211.929c-2.734,2.734-2.734,7.166,0,9.899l0,0c2.734,2.734,7.166,2.734,9.899,0c2.734-2.734,2.734-7.166,0-9.899'/%3E %3C/g%3E %3C/g%3E %3C/g%3E %3C/svg%3E");
}
@keyframes reactistRotateRight {
  from {
    transform: rotate(0deg);
    transform-origin: center center;
  }
  to {
    transform: rotate(360deg);
    transform-origin: center center;
  }
}

.reactist_tooltip__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_tooltip__wrapper .reactist_tooltip__text {
  font-size: 0.81rem;
  color: #202020;
  font-weight: 400;
  line-height: 1.6;
  text-align: center;
  text-overflow: ellipsis;
  overflow: hidden;
  max-width: 100%;
  padding: 5px 10px;
  max-width: 250px;
}
.reactist_tooltip__wrapper .reactist_tooltip__text:not(.inverted) {
  background-color: #333333;
  color: white;
  border: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:before {
  border-bottom-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_top:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:before {
  border-left-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_right:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:before {
  border-top-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_bottom:after {
  display: none;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:before {
  border-right-color: #333333;
}
.reactist_tooltip__wrapper .reactist_tooltip__arrow:not(.inverted).arrow_left:after {
  display: none;
}

.reactist_popover {
  white-space: normal;
  z-index: 1000;
  position: fixed;
  top: 0;
  left: 0;
  display: none;
  visibility: hidden;
  opacity: 0;
  transition: opacity 0.1s linear;
}
.reactist_popover__wrapper {
  display: inline-block;
  max-width: 100%;
  cursor: unset;
}
.reactist_popover .reactist_popover__content {
  max-width: 100%;
  display: inline-flex;
  border-radius: 3px;
  color: #202020;
  background-color: white;
  border: 1px solid #dcdcdc;
  box-shadow: 0 1px 8px 0 rgba(0, 0, 0, 0.08);
}
.reactist_popover.visible {
  display: inline-block;
  display: initial;
  visibility: visible;
  opacity: 1;
}
.reactist_popover.arrow_top:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid #dcdcdc;
}
.reactist_popover.arrow_top:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  top: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-bottom: 5px solid white;
}
.reactist_popover.arrow_right:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -5px;
  border-left: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_right:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  right: -4px;
  border-left: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_bottom:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -5px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid #dcdcdc;
}
.reactist_popover.arrow_bottom:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: -4px;
  left: calc(50% - 5px);
  border-left: 5px solid transparent;
  border-right: 5px solid transparent;
  border-top: 5px solid white;
}
.reactist_popover.arrow_left:before {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -5px;
  border-right: 5px solid #dcdcdc;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}
.reactist_popover.arrow_left:after {
  position: absolute;
  content: '';
  height: 0;
  width: 0;
  bottom: calc(50% - 5px);
  left: -4px;
  border-right: 5px solid white;
  border-bottom: 5px solid transparent;
  border-top: 5px solid transparent;
}

