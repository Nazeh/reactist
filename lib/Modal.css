@keyframes fadein {
  from {
    opacity: 0;
  }
  to {
    opacity: 1;
  }
}
@keyframes spinner {
  0% {
    transform: rotate(0deg);
  }
  100% {
    transform: rotate(360deg);
  }
}
.reactist_overlay {
  position: fixed;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  z-index: 2;
  display: flex;
  justify-content: center;
  background: rgba(0, 0, 0, 0.7);
  overflow: auto;
  animation: fadein 0.25s;
}
.reactist_overlay_inner {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin: 40px 0;
  width: 100%;
}
.reactist_modal_box {
  display: flex;
  flex-direction: column;
  overflow: auto;
  overflow-x: hidden;
  width: 580px;
  border-radius: 3px;
  box-shadow: 0px 2px 8px 0px rgba(0, 0, 0, 0.16);
}
.reactist_modal_box.medium {
  width: 680px;
}
.reactist_modal_box.large {
  width: 60%;
  max-width: 1000px;
}
@media only screen and (max-width: 992px) {
  .reactist_modal_box.large {
    width: calc(100% - 80px);
  }
}
@media only screen and (max-width: 992px) {
  .reactist_modal_box {
    width: calc(100% - 80px);
    max-width: 580px;
  }
}
.reactist_modal_box__header {
  display: flex;
  flex-shrink: 0;
  align-items: center;
  background-color: #f9f9f9;
  border-top-left-radius: 4px;
  border-top-right-radius: 4px;
  border-bottom: 1px solid #fafafa;
}
.reactist_modal_box__header p {
  margin: 0 0 0 25px;
  padding: 10px 0;
  font-size: 0.875rem;
  color: #202020;
  font-weight: 700;
  line-height: 1.7;
}
.reactist_modal_box__header .title {
  font-size: 1rem;
  color: #202020;
  font-weight: 700;
  line-height: 1.8;
  display: block;
  margin-top: 5px;
}
.reactist_modal_box__header .subtitle {
  font-size: 0.81rem;
  color: #808080;
  font-weight: 400;
  line-height: 1.6;
  display: block;
  margin-top: 5px;
}
.reactist_modal_box__header a {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 46px;
  height: 46px;
  margin-left: auto;
  margin-right: 8px;
}
.reactist_modal_box__body {
  flex-grow: 1;
  overflow: auto;
  padding: 20px 25px 25px 25px;
  background-color: white;
  font-size: 0.875rem;
  color: #202020;
  font-weight: 400;
  line-height: 1.7;
}
.reactist_modal_box__body.plain {
  padding: 0;
}
.reactist_modal_box__body .dialog {
  overflow: auto;
}
.reactist_modal_box__body .dialog .reactist_icon {
  width: 42px;
  float: left;
  margin-right: 20px;
}
.reactist_modal_box__body .dialog .reactist_icon > * {
  max-height: 42px;
  max-width: 100%;
}
.reactist_modal_box__body .dialog .content {
  float: right;
  width: calc(100% - 62px);
}
.reactist_modal_box__body .dialog .content h1 {
  margin: 0;
  font-size: 0.875rem;
  color: #202020;
  font-weight: 700;
  line-height: 1.7;
  line-height: 1;
}
.reactist_modal_box__body:after {
  clear: both;
  content: '';
  display: block;
}
.reactist_modal_box__body .close {
  float: right;
}
.reactist_modal_box__actions {
  display: flex;
  flex-shrink: 0;
  justify-content: flex-end;
  padding: 20px 25px;
  background-color: white;
  border-top: 1px solid #ececec;
}
.reactist_modal_box__actions button {
  margin-left: 10px;
}

