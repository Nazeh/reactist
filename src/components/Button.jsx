import './styles/button.less'

import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import Tooltip from './Tooltip'

class Button extends React.Component {
    constructor(props, context) {
        super(props, context)
        this._onClick = this._onClick.bind(this)
    }

    _onClick(event) {
        event.preventDefault()
        if (!this.props.disabled && this.props.onClick) {
            this.props.onClick()
        }
    }

    render() {
        const { data_tip } = this.props

        const className = classNames(
            {
                reactist: true,
                button: true,
                secondary: this.props.secondary,
                small: this.props.small,
                white: this.props.white,
                loading: this.props.loading,
                danger: this.props.danger
            },
            this.props.className
        )

        const button = (
            <button
                className={className}
                disabled={this.props.disabled}
                onClick={this._onClick}
            >
                <div className="wrapper">
                    <span>{this.props.name}</span>
                </div>
            </button>
        )

        // conditionally wrap into tooltip
        return data_tip ? <Tooltip text={data_tip}>{button}</Tooltip> : button
    }
}
Button.displayName = 'Button'
Button.defaultProps = {
    secondary: false,
    small: false,
    white: false,
    loading: false,
    disabled: false,
    danger: false
}
Button.propTypes = {
    /** Text that is displayed on the button. */
    name: PropTypes.string,
    /**
     * Function that is called when the button is clicked.
     * Is only invoked when disabled is not set.
     */
    onClick: PropTypes.func,
    /** Secondary style. */
    secondary: PropTypes.bool,
    /** Small style. */
    small: PropTypes.bool,
    /** White style. */
    white: PropTypes.bool,
    /** Loading style. */
    loading: PropTypes.bool,
    /** Disabled style. Prevents onClick from being called. */
    disabled: PropTypes.bool,
    /** Danger style. */
    danger: PropTypes.bool,
    /** Tooltip that is displayed on hover. */
    data_tip: PropTypes.string,
    /** Additional css class applied to the button. */
    className: PropTypes.string
}

export default Button
