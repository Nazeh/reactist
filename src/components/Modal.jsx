import './styles/modal.less'

import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'

import CloseIcon from './icons/CloseIcon.svg'

class Box extends React.Component {
    constructor(props, context) {
        super(props, context)
        this._handleKeyDown = this._handleKeyDown.bind(this)
        this._closeModal = this._closeModal.bind(this)
        this._handleOverlayClick = this._handleOverlayClick.bind(this)
        window.addEventListener('keydown', this._handleKeyDown)
        window.addEventListener('click', this._handleOverlayClick)
    }

    componentWillUnmount() {
        window.removeEventListener('keydown', this._handleKeyDown)
        window.removeEventListener('click', this._handleOverlayClick)
    }

    _closeModal() {
        ReactDOM.unmountComponentAtNode(document.getElementById('modal_box'))
    }

    _handleKeyDown(event) {
        if (event.keyCode === 27) {
            // ESC
            this._closeModal()
        }
    }

    _handleOverlayClick(event) {
        if (
            this.props.closeOnOverlayClick &&
            event &&
            event.target &&
            event.target.id === 'reactist-overlay'
        ) {
            this._closeModal()
        }
    }

    render() {
        let class_name = 'reactist_modal_box'

        if (this.props.large) {
            class_name += ' large'
        }

        if (this.props.className) {
            class_name += ` ${this.props.className}`
        }

        return (
            <div className="reactist_overlay" id="reactist-overlay">
                <div className={class_name}>{this.props.children}</div>
            </div>
        )
    }
}
Box.displayName = 'Modal.Box'
Box.defaultProps = {
    large: false,
    closeOnOverlayClick: false
}
Box.propTypes = {
    /** Additional css class applied to the Modal.Box. */
    className: PropTypes.string,
    /** Large style. */
    large: PropTypes.bool,
    /** Close the Modal when clicking on the overlay. */
    closeOnOverlayClick: PropTypes.bool,
    /** Children to render inside the Modal.Box. Normally Modal.Header, Modal.Body and Modal.Actions. */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
}

class Header extends React.Component {
    _closeModal(event) {
        event.preventDefault()
        if (typeof this.props.beforeClose === 'function') {
            this.props.beforeClose()
        }
        ReactDOM.unmountComponentAtNode(document.getElementById('modal_box'))
    }

    render() {
        return (
            <div className="reactist_modal_box__header">
                <p>
                    {this.props.title && (
                        <span className="title">{this.props.title}</span>
                    )}
                    {this.props.subtitle && (
                        <span className="subtitle">{this.props.subtitle}</span>
                    )}
                    {this.props.children}
                </p>
                <a
                    className="close"
                    onClick={this._closeModal.bind(this)}
                    href="#"
                >
                    <CloseIcon />
                </a>
            </div>
        )
    }
}
Header.displayName = 'Modal.Header'
Header.propTypes = {
    /** Children to render inside the Modal.Header for a fully customizable appearance. */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ]),
    /** Title of the Modal.Header. */
    title: PropTypes.string,
    /** Subtitle of the Modal.Header. */
    subtitle: PropTypes.string,
    /** Function that is called right before the Modal unmounts itself. */
    beforeClose: PropTypes.func
}

class Body extends React.Component {
    _closeModal(event) {
        event.preventDefault()
        ReactDOM.unmountComponentAtNode(document.getElementById('modal_box'))
    }

    render() {
        let className = 'reactist_modal_box__body'
        if (this.props.className) {
            className += ` ${this.props.className}`
        }

        return (
            <div className={className}>
                {this.props.showCloseIcon && (
                    <a
                        className="close"
                        onClick={this._closeModal.bind(this)}
                        href="#"
                    >
                        <CloseIcon />
                    </a>
                )}
                {this.props.icon ? (
                    <div className="dialog">
                        <div className="icon">{this.props.icon}</div>
                        <div className="content">{this.props.children}</div>
                    </div>
                ) : (
                    this.props.children
                )}
            </div>
        )
    }
}
Body.displayName = 'Modal.Body'
Body.defaultProps = {
    showCloseIcon: false
}
Body.propTypes = {
    /** Display an icon (or basically any component) on the right hand side of the Modal.Body. */
    icon: PropTypes.node,
    /**
     * Render a close icon in the top right corner of the Modal.Body.
     * Recommended to use when no Modal.Header is used.
     */
    showCloseIcon: PropTypes.bool,
    /** Additionall css class applied to the Modal.Body. */
    className: PropTypes.string,
    /** Children to render inside the Modal.Body. */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
}

class Actions extends React.Component {
    _onClick(on_click) {
        if (typeof on_click === 'function') {
            on_click()
        }
        ReactDOM.unmountComponentAtNode(document.getElementById('modal_box'))
    }

    render() {
        const children = React.Children.map(this.props.children, child => {
            if (!child) return false
            if (child.props.close) {
                return React.cloneElement(child, {
                    onClick: () => this._onClick(child.props.onClick)
                })
            } else {
                return React.cloneElement(child)
            }
        })

        return <div className="reactist_modal_box__actions">{children}</div>
    }
}
Actions.displayName = 'Modal.Actions'
Actions.propTypes = {
    /**
     * Children to render inside the Modal.Actions. They can have an optional `close` property (boolean).
     * When that is supplied and set to true it will close the modal after the onClick function
     */
    children: PropTypes.oneOfType([
        PropTypes.arrayOf(PropTypes.node),
        PropTypes.node
    ])
}

export default {
    Box,
    Header,
    Body,
    Actions
}
