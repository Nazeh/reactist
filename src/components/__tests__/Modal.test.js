jest.mock('react-dom')

import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'

import Modal from '../Modal'
import Button from '../Button' // for more descriptive snapshots

describe('Modal.Box', () => {
    it('renders without crashing', () => {
        const box = shallow(<Modal.Box />)
        expect(toJson(box)).toMatchSnapshot()
    })

    it('renders all children without crashing', () => {
        const box = shallow(
            <Modal.Box>
                <Modal.Header>Realistic test</Modal.Header>
                <Modal.Body>... is very realistic</Modal.Body>
                <Modal.Actions>
                    <Button name="Wow" />
                    <Button name="So real" />
                </Modal.Actions>
            </Modal.Box>
        )
        expect(toJson(box)).toMatchSnapshot()
    })

    it('adds additionally supplied className prop', () => {
        const box = shallow(<Modal.Box className="applied-to-child-of-box" />)
        expect(toJson(box)).toMatchSnapshot(box)
    })

    it('adds large to className when supplied', () => {
        const box = shallow(<Modal.Box large />)
        expect(toJson(box)).toMatchSnapshot()
    })

    it('unmounts the modal_box when escape was pressed', () => {
        const box = mount(<Modal.Box />).instance()

        const preventSpy = jest.fn()
        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        box._handleKeyDown({ keyCode: 27, preventDefault: preventSpy })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
        expect(getCallCount(preventSpy)).toBe(1)
    })

    it('does nothing when any other key than escape was pressed', () => {
        const box = mount(<Modal.Box />).instance()

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        box._handleKeyDown({ keyCode: 23 })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount
        )
    })

    it('unmounts the modal_box when clicking on overlay', () => {
        const box = mount(<Modal.Box closeOnOverlayClick />).instance()

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        box._handleOverlayClick({ target: { id: 'reactist-overlay' } })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })

    it('unmounts the modal_box when clicking on inner overlay (sides of modal)', () => {
        const box = mount(<Modal.Box closeOnOverlayClick />).instance()

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        box._handleOverlayClick({ target: { id: 'reactist-overlay-inner' } })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })

    it('does not close on overlay click when property was not supplied', () => {
        const box = mount(<Modal.Box />).instance()

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        box._handleOverlayClick({ target: { id: 'reactist-overlay' } })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount
        )
    })
})

describe('Modal.Header', () => {
    it('renders without crashing', () => {
        const header = shallow(<Modal.Header />)
        expect(toJson(header)).toMatchSnapshot()
    })

    it('renders title and subtitle if available', () => {
        const headerTitle = shallow(<Modal.Header title="Modal Title" />)
        expect(toJson(headerTitle)).toMatchSnapshot()

        const headerSubtitle = shallow(
            <Modal.Header subtitle="Modal Subtitle" />
        )
        expect(toJson(headerSubtitle)).toMatchSnapshot()
    })

    it('renders all children without crashing', () => {
        const textHeader = shallow(
            <Modal.Header>Just a Text Header</Modal.Header>
        )
        expect(toJson(textHeader)).toMatchSnapshot()

        const complexHeader = shallow(
            <Modal.Header>
                <Button name="Wow this looks complex" />
                Some additional text as well
            </Modal.Header>
        )
        expect(toJson(complexHeader)).toMatchSnapshot()
    })

    it('clicking the close icon unmounts the modal_box', () => {
        const header = shallow(<Modal.Header />)

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        header.find('.close').simulate('click', { preventDefault: jest.fn() })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })

    it('invokes beforeClose when closing the modal', () => {
        const beforeCloseSpy = jest.fn()
        const header = shallow(<Modal.Header beforeClose={beforeCloseSpy} />)

        header.find('.close').simulate('click', { preventDefault: jest.fn() })
        expect(beforeCloseSpy).toHaveBeenCalled()
    })
})

describe('Modal.Body', () => {
    it('renders without crashing', () => {
        const body = shallow(<Modal.Body />)
        expect(toJson(body)).toMatchSnapshot()
    })

    it('renders all children without crashing', () => {
        const body = shallow(
            <Modal.Body>
                <div>
                    <h1>This is a body</h1>
                </div>
                <Button name="Button in Body" />
            </Modal.Body>
        )
        expect(toJson(body)).toMatchSnapshot()
    })

    it('renders an icon when supplied via props', () => {
        const body = shallow(
            <Modal.Body icon={<img src="awesome/image.png" />} />
        )
        expect(toJson(body)).toMatchSnapshot()
    })

    it('clicking the close icon unmounts the modal_box', () => {
        const body = shallow(<Modal.Body showCloseIcon />)

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        body.find('.close').simulate('click', { preventDefault: jest.fn() })
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })

    it('adds additionally supplied className prop', () => {
        const body = shallow(<Modal.Body className="applied-to-modal-body" />)
        expect(toJson(body)).toMatchSnapshot()
    })
})

describe('Modal.Actions', () => {
    it('renders without crashing', () => {
        const actions = shallow(<Modal.Actions />)
        expect(toJson(actions)).toMatchSnapshot()
    })

    it('renders all children without crashing', () => {
        const actions = shallow(
            <Modal.Actions>
                <Button name="Action 1" />
                <Button name="Action 2" />
            </Modal.Actions>
        )
        expect(toJson(actions)).toMatchSnapshot()
    })

    it('unmounts modal_box when child with close prop is clicked', () => {
        const clickSpy = jest.fn()
        const actions = shallow(
            <Modal.Actions>
                <Button name="Action 1" close onClick={clickSpy} />
                <Button name="Action 2" />
            </Modal.Actions>
        )

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        actions
            .find('Button')
            .first()
            .simulate('click')
        expect(clickSpy).toHaveBeenCalled()
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })

    it('unmounts modal_box when child with close prop is clicked even when no onClick is specified', () => {
        const actions = shallow(
            <Modal.Actions>
                <Button name="Action 1" close />
            </Modal.Actions>
        )

        const unmountCallCount = getCallCount(ReactDOM.unmountComponentAtNode)
        actions
            .find('Button')
            .first()
            .simulate('click')
        expect(getCallCount(ReactDOM.unmountComponentAtNode)).toBe(
            unmountCallCount + 1
        )
    })
})

// Helper functions ///////////////////////////////////////////////////////////
function getCallCount(mock) {
    return mock.mock.calls.length
}
