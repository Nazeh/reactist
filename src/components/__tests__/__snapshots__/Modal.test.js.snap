// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal.Actions renders all children without crashing 1`] = `
<div
  className="reactist_modal_box__actions"
>
  <Button
    danger={false}
    disabled={false}
    key=".0"
    loading={false}
    name="Action 1"
    secondary={false}
    small={false}
    white={false}
  />
  <Button
    danger={false}
    disabled={false}
    key=".1"
    loading={false}
    name="Action 2"
    secondary={false}
    small={false}
    white={false}
  />
</div>
`;

exports[`Modal.Actions renders without crashing 1`] = `
<div
  className="reactist_modal_box__actions"
/>
`;

exports[`Modal.Body adds additionally supplied className prop 1`] = `
<div
  className="reactist_modal_box__body applied-to-modal-body"
/>
`;

exports[`Modal.Body renders all children without crashing 1`] = `
<div
  className="reactist_modal_box__body"
>
  <div>
    <h1>
      This is a body
    </h1>
  </div>
  <Button
    danger={false}
    disabled={false}
    loading={false}
    name="Button in Body"
    secondary={false}
    small={false}
    white={false}
  />
</div>
`;

exports[`Modal.Body renders an icon when supplied via props 1`] = `
<div
  className="reactist_modal_box__body"
>
  <div
    className="dialog"
  >
    <div
      className="icon"
    >
      <img
        src="awesome/image.png"
      />
    </div>
    <div
      className="content"
    />
  </div>
</div>
`;

exports[`Modal.Body renders without crashing 1`] = `
<div
  className="reactist_modal_box__body"
/>
`;

exports[`Modal.Box adds additionally supplied className prop: [object Object] 1`] = `
<div
  className="reactist_overlay"
  id="reactist-overlay"
>
  <div
    className="reactist_overlay_inner"
    id="reactist-overlay-inner"
  >
    <div
      className="reactist_modal_box applied-to-child-of-box"
    />
  </div>
</div>
`;

exports[`Modal.Box adds large to className when supplied 1`] = `
<div
  className="reactist_overlay"
  id="reactist-overlay"
>
  <div
    className="reactist_overlay_inner"
    id="reactist-overlay-inner"
  >
    <div
      className="reactist_modal_box large"
    />
  </div>
</div>
`;

exports[`Modal.Box renders all children without crashing 1`] = `
<div
  className="reactist_overlay"
  id="reactist-overlay"
>
  <div
    className="reactist_overlay_inner"
    id="reactist-overlay-inner"
  >
    <div
      className="reactist_modal_box"
    >
      <Modal.Header>
        Realistic test
      </Modal.Header>
      <Modal.Body
        showCloseIcon={false}
      >
        ... is very realistic
      </Modal.Body>
      <Modal.Actions>
        <Button
          danger={false}
          disabled={false}
          loading={false}
          name="Wow"
          secondary={false}
          small={false}
          white={false}
        />
        <Button
          danger={false}
          disabled={false}
          loading={false}
          name="So real"
          secondary={false}
          small={false}
          white={false}
        />
      </Modal.Actions>
    </div>
  </div>
</div>
`;

exports[`Modal.Box renders without crashing 1`] = `
<div
  className="reactist_overlay"
  id="reactist-overlay"
>
  <div
    className="reactist_overlay_inner"
    id="reactist-overlay-inner"
  >
    <div
      className="reactist_modal_box"
    />
  </div>
</div>
`;

exports[`Modal.Header renders all children without crashing 1`] = `
<div
  className="reactist_modal_box__header"
>
  <p>
    Just a Text Header
  </p>
  <a
    className="close"
    href="#"
    onClick={[Function]}
  >
    <CloseIcon />
  </a>
</div>
`;

exports[`Modal.Header renders all children without crashing 2`] = `
<div
  className="reactist_modal_box__header"
>
  <p>
    <Button
      danger={false}
      disabled={false}
      loading={false}
      name="Wow this looks complex"
      secondary={false}
      small={false}
      white={false}
    />
    Some additional text as well
  </p>
  <a
    className="close"
    href="#"
    onClick={[Function]}
  >
    <CloseIcon />
  </a>
</div>
`;

exports[`Modal.Header renders title and subtitle if available 1`] = `
<div
  className="reactist_modal_box__header"
>
  <p>
    <span
      className="title"
    >
      Modal Title
    </span>
  </p>
  <a
    className="close"
    href="#"
    onClick={[Function]}
  >
    <CloseIcon />
  </a>
</div>
`;

exports[`Modal.Header renders title and subtitle if available 2`] = `
<div
  className="reactist_modal_box__header"
>
  <p>
    <span
      className="subtitle"
    >
      Modal Subtitle
    </span>
  </p>
  <a
    className="close"
    href="#"
    onClick={[Function]}
  >
    <CloseIcon />
  </a>
</div>
`;

exports[`Modal.Header renders without crashing 1`] = `
<div
  className="reactist_modal_box__header"
>
  <p />
  <a
    className="close"
    href="#"
    onClick={[Function]}
  >
    <CloseIcon />
  </a>
</div>
`;
